src/theory/quantifiers/sygus/sygus_enumerator.cpp

```diff
@@ -0,0 +1,924 @@
+        // more aggressive merging of constructor classes. On the negative side,
+        // this adds another level of indirection to remember which argument
+        // positions the argument types occur in, for each constructor.
+        Node n = nm->mkConst(Rational(i));
+        nToC[n] = i;
+        tnit.add(n, argTypes[i]);
+      }
+    }
+    std::map<Node, unsigned> assign;
+    tnit.assignIds(assign, d_numConClasses);
+    for (std::pair<const Node, unsigned>& cp : assign)
+    {
+      // determine which constructor class this goes into using tnit
+      unsigned cclassi = cp.second;
+      unsigned i = nToC[cp.first];
+      Trace("sygus-enum-debug") << "Constructor class for "
+                                << dt[i].getSygusOp() << " is " << cclassi
+                                << std::endl;
+      // initialize the constructor class
+      if (d_ccToWeight.find(cclassi) == d_ccToWeight.end())
+      {
+        d_ccToWeight[cclassi] = w;
+        d_ccToTypes[cclassi].insert(
+            d_ccToTypes[cclassi].end(), argTypes[i].begin(), argTypes[i].end());
+      }
+      // add to constructor class
+      d_ccToCons[cclassi].push_back(i);
+    }
+    Trace("sygus-enum-debug") << "#cons classes for weight <= " << w << " : "
+                              << d_numConClasses << std::endl;
+    d_weightToCcIndex[w] = d_numConClasses;
+  }
+  Trace("sygus-enum-debug") << "...finish" << std::endl;
+}
+
+unsigned SygusEnumerator::TermCache::getLastConstructorClassIndexForWeight(
+    unsigned w) const
+{
+  std::map<unsigned, unsigned>::const_iterator it = d_weightToCcIndex.find(w);
+  if (it == d_weightToCcIndex.end())
+  {
+    return d_numConClasses;
+  }
+  return it->second;
+}
+unsigned SygusEnumerator::TermCache::getNumConstructorClasses() const
+{
+  return d_numConClasses;
+}
+void SygusEnumerator::TermCache::getConstructorClass(
+    unsigned i, std::vector<unsigned>& cclass) const
+{
+  std::map<unsigned, std::vector<unsigned>>::const_iterator it =
+      d_ccToCons.find(i);
+  Assert(it != d_ccToCons.end());
+  cclass.insert(cclass.end(), it->second.begin(), it->second.end());
+}
+void SygusEnumerator::TermCache::getTypesForConstructorClass(
+    unsigned i, std::vector<TypeNode>& types) const
+{
+  std::map<unsigned, std::vector<TypeNode>>::const_iterator it =
+      d_ccToTypes.find(i);
+  Assert(it != d_ccToTypes.end());
+  types.insert(types.end(), it->second.begin(), it->second.end());
+}
+
+unsigned SygusEnumerator::TermCache::getWeightForConstructorClass(
+    unsigned i) const
+{
+  std::map<unsigned, unsigned>::const_iterator it = d_ccToWeight.find(i);
+  Assert(it != d_ccToWeight.end());
+  return it->second;
+}
+
+bool SygusEnumerator::TermCache::addTerm(Node n)
+{
+  if (!d_isSygusType)
+  {
+    Trace("sygus-enum-terms") << "tc(" << d_tn << "): term (builtin): " << n
+                              << std::endl;
+    d_terms.push_back(n);
```

Comment:
add a comment as to why it's not necessary to bother with uniqueness here

Reply:
Done.